### ISSUE TITLE
java.lang.RuntimeException: TwoVarDynamicScope does not support scopes with three or more variables

### ISSUE THREAD

**Issue by steve-rodriguez** [jruby/jruby#555]
We are seeing this issue in our JRuby application ( doesn't occur all the time, but see it quite often ).  Please advise.  Thanks!

Running in TCServer 2.6.2 sitting on top of Tomcat 7.

Configuration:  
JRuby 1.6.7
Rails 3.2.11
Jruby-Rack 1.1.6
Java JDK 1.6_0_20 ( Sun )

Error ( Stack Trace which doesn't point to any specific line of code that we developed ):

```
26-Feb-2013 13:39:57.012 SEVERE org.apache.catalina.core.ApplicationContext.log Application Error
 java.lang.RuntimeException: TwoVarDynamicScope does not support scopes with three or more variables
        at org.jruby.runtime.scope.NoVarsDynamicScope.setValueTwoDepthZero(NoVarsDynamicScope.java:100)
        at rubyjit.ActionDispatch::ShowExceptions#call_110B2FA6920293870F201A77F9D6935443A55BC3.chained_1_rescue_line_56(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/show_exceptions.rb)
        at rubyjit.ActionDispatch::ShowExceptions#call_110B2FA6920293870F201A77F9D6935443A55BC3.chained_0_rescue_1$RUBY$SYNTHETIC__file__(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/show_exceptions.rb:56)
        at rubyjit.ActionDispatch::ShowExceptions#call_110B2FA6920293870F201A77F9D6935443A55BC3.__file__(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/show_exceptions.rb:55)
        at rubyjit.ActionDispatch::ShowExceptions#call_110B2FA6920293870F201A77F9D6935443A55BC3.__file__(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/show_exceptions.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.Rails::Rack::Logger#call_app_DE29136E502379175F0E5EF42821B5467E8A6D03.chained_0_ensure_1$RUBY$__ensure__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/rack/logger.rb:32)
        at rubyjit.Rails::Rack::Logger#call_app_DE29136E502379175F0E5EF42821B5467E8A6D03.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/rack/logger.rb)
        at rubyjit.Rails::Rack::Logger#call_app_DE29136E502379175F0E5EF42821B5467E8A6D03.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/rack/logger.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:207)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:201)
        at rubyjit.Rails::Rack::Logger#call_9C3389E7CE3CCD341068E5D0C9F6283F77F6F971.block_0$RUBY$__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/rack/logger.rb:16)
        at rubyjit$Rails::Rack::Logger#call_9C3389E7CE3CCD341068E5D0C9F6283F77F6F971$block_0$RUBY$__file__.call(rubyjit$Rails::Rack::Logger#call_9C3389E7CE3CCD341068E5D0C9F6283F77F6F971$block_0$RUBY$__file__:65535)
        at org.jruby.runtime.CompiledBlock19.yield(CompiledBlock19.java:139)
        at org.jruby.runtime.CompiledBlock19.yieldSpecific(CompiledBlock19.java:101)
        at org.jruby.runtime.Block.yieldSpecific(Block.java:108)
        at rubyjit.ActiveSupport::TaggedLogging#tagged_2790A33722B3CBEBECECCFF90F769EB0F50B6FF2.chained_0_ensure_1$RUBY$__ensure__(/webapps/apps/cms/WEB-INF/gems/gems/activesupport-3.2.11/lib/active_support/tagged_logging.rb:22)
        at rubyjit.ActiveSupport::TaggedLogging#tagged_2790A33722B3CBEBECECCFF90F769EB0F50B6FF2.__file__(/webapps/apps/cms/WEB-INF/gems/gems/activesupport-3.2.11/lib/active_support/tagged_logging.rb)
        at rubyjit.ActiveSupport::TaggedLogging#tagged_2790A33722B3CBEBECECCFF90F769EB0F50B6FF2.__file__(/webapps/apps/cms/WEB-INF/gems/gems/activesupport-3.2.11/lib/active_support/tagged_logging.rb)
        at org.jruby.ast.executable.AbstractScript.__file__(AbstractScript.java:42)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:187)
        at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:176)
        at org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:187)
        at rubyjit.Rails::Rack::Logger#call_9C3389E7CE3CCD341068E5D0C9F6283F77F6F971.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/rack/logger.rb:16)
        at rubyjit.Rails::Rack::Logger#call_9C3389E7CE3CCD341068E5D0C9F6283F77F6F971.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/rack/logger.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.ActionDispatch::RequestId#call_DAE470BC0ED487EDB8E07D6A7E2E5070C41105E3.__file__(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/request_id.rb:22)
        at rubyjit.ActionDispatch::RequestId#call_DAE470BC0ED487EDB8E07D6A7E2E5070C41105E3.__file__(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/request_id.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.Rack::MethodOverride#call_5C4513A64D983C8019F0C2940E049E8355465D21.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-1.4.4/lib/rack/methodoverride.rb:21)
        at rubyjit.Rack::MethodOverride#call_5C4513A64D983C8019F0C2940E049E8355465D21.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-1.4.4/lib/rack/methodoverride.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.Rack::Runtime#call_244B79C785429BA6E33FD2DDEFEAE75AD1BF3397.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-1.4.4/lib/rack/runtime.rb:17)
        at rubyjit.Rack::Runtime#call_244B79C785429BA6E33FD2DDEFEAE75AD1BF3397.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-1.4.4/lib/rack/runtime.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.ActiveSupport::Cache::Strategy::LocalCache::Middleware#call_DD20847470D950049CFFAFD5B54D4674A615895F.chained_0_ensure_1$RUBY$__ensure__(/webapps/apps/cms/WEB-INF/gems/gems/activesupport-3.2.11/lib/active_support/cache/strategy/local_cache.rb:72)
        at rubyjit.ActiveSupport::Cache::Strategy::LocalCache::Middleware#call_DD20847470D950049CFFAFD5B54D4674A615895F.__file__(/webapps/apps/cms/WEB-INF/gems/gems/activesupport-3.2.11/lib/active_support/cache/strategy/local_cache.rb)
        at rubyjit.ActiveSupport::Cache::Strategy::LocalCache::Middleware#call_DD20847470D950049CFFAFD5B54D4674A615895F.__file__(/webapps/apps/cms/WEB-INF/gems/gems/activesupport-3.2.11/lib/active_support/cache/strategy/local_cache.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.ActionDispatch::Static#call_BA5F3A94FB90D3672B7966DC16A2150F2807BE9F.__file__(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/static.rb:62)
        at rubyjit.ActionDispatch::Static#call_BA5F3A94FB90D3672B7966DC16A2150F2807BE9F.__file__(/webapps/apps/cms/WEB-INF/gems/gems/actionpack-3.2.11/lib/action_dispatch/middleware/static.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.Rack::Cache::Context#forward_8AD88A0B9B30770151418566F2AA32B59A46FADE.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb:136)
        at rubyjit.Rack::Cache::Context#forward_8AD88A0B9B30770151418566F2AA32B59A46FADE.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:127)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:133)
        at rubyjit.Rack::Cache::Context#fetch_0AA9F2D322A01DBC085AFC27C791E82922ED9126.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb:245)
        at rubyjit.Rack::Cache::Context#fetch_0AA9F2D322A01DBC085AFC27C791E82922ED9126.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:127)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:133)
        at rubyjit.Rack::Cache::Context#lookup_398DA8272DB5A9C2C301BFE0823CABBC982DCB29.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb:185)
        at rubyjit.Rack::Cache::Context#lookup_398DA8272DB5A9C2C301BFE0823CABBC982DCB29.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:127)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:133)
        at rubyjit.Rack::Cache::Context#call!_08C06A9CB7E99EA5F657BE81F487AADFCD8CF713.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb:66)
        at rubyjit.Rack::Cache::Context#call!_08C06A9CB7E99EA5F657BE81F487AADFCD8CF713.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.Rack::Cache::Context#call_547F7BF5B79EE8A794CE388DDE6EB7E2CDCE78D2.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb:51)
        at rubyjit.Rack::Cache::Context#call_547F7BF5B79EE8A794CE388DDE6EB7E2CDCE78D2.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at rubyjit.Rack::Cache::Context#call_547F7BF5B79EE8A794CE388DDE6EB7E2CDCE78D2.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb:51)
        at rubyjit.Rack::Cache::Context#call_547F7BF5B79EE8A794CE388DDE6EB7E2CDCE78D2.__file__(/webapps/apps/cms/WEB-INF/gems/gems/rack-cache-1.2/lib/rack/cache/context.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.Rails::Engine#call_6171CBECE68C5D2D005B760B76FF27F1A65333DC.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/engine.rb:479)
        at rubyjit.Rails::Engine#call_6171CBECE68C5D2D005B760B76FF27F1A65333DC.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/engine.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:187)
        at org.jruby.runtime.callsite.SuperCallSite.callBlock(SuperCallSite.java:190)
        at org.jruby.runtime.callsite.SuperCallSite.call(SuperCallSite.java:197)
        at rubyjit.Rails::Application#call_69A91512DA2942A380BCE125003291ED53C13C8E.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/application.rb:223)
        at rubyjit.Rails::Application#call_69A91512DA2942A380BCE125003291ED53C13C8E.__file__(/webapps/apps/cms/WEB-INF/gems/gems/railties-3.2.11/lib/rails/application.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:167)
        at rubyjit.Rack::Handler::Servlet#call_7F5B7C69190ACC22B41B2797381AFBF1F7A26DDC.__file__(file:/webapps/apps/cms/WEB-INF/lib/jruby-rack-1.1.4.jar!/rack/handler/servlet.rb:20)
        at rubyjit.Rack::Handler::Servlet#call_7F5B7C69190ACC22B41B2797381AFBF1F7A26DDC.__file__(file:/webapps/apps/cms/WEB-INF/lib/jruby-rack-1.1.4.jar!/rack/handler/servlet.rb)
        at org.jruby.internal.runtime.methods.JittedMethod.call(JittedMethod.java:167)
        at org.jruby.RubyClass.finvoke(RubyClass.java:699)
        at org.jruby.javasupport.util.RuntimeHelpers.invoke(RuntimeHelpers.java:548)
        at org.jruby.RubyBasicObject.callMethod(RubyBasicObject.java:361)
        at org.jruby.javasupport.JavaEmbedUtils$1.callMethod(JavaEmbedUtils.java:143)
        at org.jruby.rack.DefaultRackApplication.__call(DefaultRackApplication.java:72)
        at org.jruby.rack.DefaultRackApplication.call(DefaultRackApplication.java:40)
        at org.jruby.rack.AbstractRackDispatcher.process(AbstractRackDispatcher.java:30)
        at org.jruby.rack.AbstractFilter.doFilter(AbstractFilter.java:41)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:243)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:210)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:224)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:169)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:472)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:168)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:100)
        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:929)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:118)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:405)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:964)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:515)
        at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:304)
        at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
        at java.lang.Thread.run(Thread.java:619)
```

**Comment by headius** [comment]
Issues very similar to this were fixed later in the 1.7 series, and the 9k series has a completely new runtime that makes this bug somewhat irrelevant, so let's proactively close this.

@digitalextremist If you can still reproduce on some recent JRuby, we can reopen and get it fixed up.

**Comment by digitalextremist** [comment]
Wow! Serious throw-back action. `2013` is `in the house` -- thanks @headius, I am long gone here.
